src/main/scala/com/fulcrumgenomics/bam/Bams.scala

```diff
@@ -490,4 +492,90 @@ object Bams extends LazyLogging {
+    * @param outOrder the desired output sort order, if any
+    * @param ref the path to the reference FASTA, used for CRAM output
+    * @param maxRecordsInMemoryWhenSorting the maximum number of records to keep in memory when sorting
+    */
+  def writerFromTemplateIterator(output: PathToBam,
+                                 inHeader: SAMFileHeader,
+                                 outOrder: Option[SamOrder] = None,
+                                 ref: Option[PathToFasta] = None,
+                                 maxRecordsInMemoryWhenSorting: Int = 250e3.toInt
+                                ): SamWriter = {
+    val outHeader    = inHeader.clone()
+    val inSortOrder  = inHeader.getSortOrder
+    val inGroupOrder = inHeader.getGroupOrder
+    val inSubSort    = Option(inHeader.getAttribute("SS"))
+
+    // Get the sort order, if any, to force the writer to resort.
+    val writerSortOrder = {
+      val canSkipSortingInput = inSortOrder == SortOrder.queryname || inGroupOrder == GroupOrder.query
+      (outOrder, canSkipSortingInput) match {
+        case (None,        false) =>
+          // already sorted to Queryname after filtering, so the output will be Queryname, and so no need to re-sort the output
+          SamOrder.Queryname.applyTo(outHeader)
+          None
+        case (Some(order), false) =>
+          // already sorted to Queryname after filtering, so don't sort the output if the given output sort order matches Queryname
+          order.applyTo(outHeader)
+          if (order != SamOrder.Queryname) Some(order) else None
+        case (None,         true) =>
+          // input did not need to be re-sorted, so do not re-sort the output
+          None
+        case (Some(order), true)  =>
+          // If sorting prior to filtering was not required, then the input was queryname or query grouped.  So sort the
+          // output only if the specified output `sortOrder` does not match the input sort order.
+          if (order.sortOrder == inSortOrder && order.groupOrder == inGroupOrder && order.subSort == inSubSort) {
+            None // input matches the output order, no need to force sorting, keep the header
+          }
+          else { // mismatch in the input order and output order, force sorting the output
+            order.applyTo(outHeader)
+            Some(order)
+          }
+      }
+    }
+
+    val writer = SamWriter(output, outHeader, sort=writerSortOrder, ref=ref, maxRecordsInRam=maxRecordsInMemoryWhenSorting)
+    writerSortOrder.foreach(o => logger.info(f"Output will be sorted into $o order"))
+    writer
+  }
+
+  /** Builds a [[Writer]] of [[SamRecord]]s that regenerates the NM, UQ, and MD tags using the given map of reference
+    * sequences.
+    *
+    * @param writer the writer to write to
+    * @param ref the path to the reference FASTA
+    */
+  def regenerateNmUqMdTagsWriter(writer: SamWriter, ref: PathToFasta): Writer[SamRecord] with Closeable = {
```

Comment:
ditto, pulled out from `FilterConsensusReads` and happy to write tests

Reply:
How about `nmUqMdTagRegeneratingWriter()`?  I think it needs to be `{adjective}Writer`.